libcache/cache.c

```diff
@@ -0,0 +1,696 @@
+{
+	int i, lineIndex;
+	uint64_t addr = 0;
+	cacheline_t *temp = NULL, *linePtr = NULL;
+	unsigned char flags = 0;
+	void *data;
+	ssize_t written = cache->lineSize;
+	cacheset_t *setPtr = &cache->sets[setIndex];
+
+	if (setPtr->count < LIBCACHE_NUM_WAYS) {
+		setPtr->count++;
+
+		for (i = 0; i < setPtr->count; ++i) {
+			if (!IS_VALID(setPtr->lines[i].flags)) {
+				lineIndex = i;
+				break;
+			}
+		}
+
+		temp = &(setPtr->lines[lineIndex]);
+
+		for (i = 0; i < setPtr->count; ++i) {
+			if (setPtr->tags[i] == NULL) {
+				setPtr->tags[i] = temp;
+				break;
+			}
+		}
+
+		LIST_ADD(&setPtr->timestamps, temp);
+	}
+	else {
+		temp = setPtr->timestamps;
+
+		if (IS_DIRTY(temp->flags)) {
+			addr = cache_computeAddr(cache, temp->tag, setIndex);
+			written = cache_flushLine(cache, temp, addr);
+		}
+
+		if (written == (ssize_t)cache->lineSize) {
+			LIST_REMOVE(&setPtr->timestamps, temp);
+			free(temp->data);
+
+			for (i = 0; i < setPtr->count; ++i) {
+				if (IS_VALID(setPtr->lines[i].flags) && (setPtr->lines[i].tag == temp->tag)) {
+					lineIndex = i;
+					break;
+				}
+			}
+
+			LIST_ADD(&setPtr->timestamps, temp);
+		}
+	}
+
+	if (written == (ssize_t)cache->lineSize) {
+		linePtr = &(setPtr->lines[lineIndex]);
+
+		data = malloc(cache->lineSize * sizeof(unsigned char));
+
+		if (data == NULL) {
+			linePtr = NULL;
+		}
+		else {
+			linePtr->data = data;
+			linePtr->tag = tag;
+			SET_VALID(flags);
+			linePtr->flags = flags;
+
+			qsort(setPtr->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+		}
+	}
+
+	return linePtr;
+}
+
+
+static cacheline_t *cache_findLine(cacheset_t *setPtr, uint64_t tag, int update)
+{
+	cacheline_t key, *keyPtr = &key, *linePtr = NULL, **temp = NULL;
+
+	key.tag = tag;
+
+	temp = bsearch(&keyPtr, setPtr->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	if (temp != NULL) {
+		if (update != 0) {
+			LIST_REMOVE(&setPtr->timestamps, *temp);
+			LIST_ADD(&setPtr->timestamps, *temp);
+		}
+
+		linePtr = *temp;
+	}
+
+	return linePtr;
+}
+
+
+static uint64_t cache_computeOffset(const cachectx_t *cache, const uint64_t addr)
+{
+	return addr & cache->offMask;
+}
+
+
+static uint64_t cache_computeSetIndex(const cachectx_t *cache, const uint64_t addr)
+{
+	return (addr >> cache->offBitsNum) & cache->setMask;
+}
+
+
+static uint64_t cache_computeTag(const cachectx_t *cache, const uint64_t addr)
+{
+	return (addr >> (cache->offBitsNum + cache->setBitsNum)) & cache->tagMask;
+}
+
+
+static size_t cache_computeTempCount(const size_t left, const size_t lineSize, uint64_t *addr, uint64_t *offset, const size_t count, const size_t remainder)
+{
+	size_t tempCount = 0;
+
+	if (left == count) {
+		if (count <= (lineSize - *offset)) {
+			tempCount = count;
+		}
+		else {
+			tempCount = lineSize - *offset;
+		}
+		*addr -= *offset;
+	}
+	else if (left == remainder) {
+		tempCount = remainder;
+		*offset = 0;
+	}
+	else {
+		tempCount = lineSize;
+		*offset = 0;
+	}
```

Comment:
Less lines, matter of taste, so just a suggestion
```suggestion
	else {
		*offset = 0;
		tempCount = (left == remainder) ? remainder : lineSize;
	}
```

Reply:
I like this shorter approach. Thank you, I'll use it.